src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -130,104 +156,133 @@ public static uint GetResponseHeaderNumberInfo(SafeWinHttpHandle requestHandle,
             return result;
         }
 
-        private static string GetResponseHeaderStringInfo(SafeWinHttpHandle requestHandle, uint infoLevel)
+        private static char[] CreateResponseHeaderInfoBuffer(SafeWinHttpHandle requestHandle, uint infoLevel)
         {
-            uint bytesNeeded = 0;
-            bool results = false;
+            uint bufferLengthInBytes = 0;
 
-            // Call WinHttpQueryHeaders once to obtain the size of the buffer needed.  The size is returned in
+            // Call WinHttpQueryHeaders to obtain the size of the buffer needed. The size is returned in
             // bytes but the API actually returns Unicode characters.
             if (!Interop.WinHttp.WinHttpQueryHeaders(
                 requestHandle,
                 infoLevel,
                 Interop.WinHttp.WINHTTP_HEADER_NAME_BY_INDEX,
-                null,
-                ref bytesNeeded,
+                IntPtr.Zero,
+                ref bufferLengthInBytes,
                 IntPtr.Zero))
             {
                 int lastError = Marshal.GetLastWin32Error();
-                if (lastError == Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND)
+
+                Debug.Assert(lastError != Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND);
+
+                if (lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER)
                 {
-                    return null;
+                    throw WinHttpException.CreateExceptionUsingError(lastError);
                 }
+            }
 
-                if (lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER)
+            return new char[bufferLengthInBytes / 2];
+        }
+
+        private static int GetResponseHeaderInfo(SafeWinHttpHandle requestHandle, uint infoLevel, char[] buffer)
+        {
+            Debug.Assert(buffer != null, "buffer must not be null.");
+            Debug.Assert(buffer.Length > 0, "buffer must not be empty.");
+
+            uint bufferLengthInBytes = (uint)buffer.Length * 2;
+
+            GCHandle pinnedHandle = GCHandle.Alloc(buffer, GCHandleType.Pinned);
```

Comment:
It'd be nice to use fixed instead of a pinned GCHandle here, as it's faster.  Are you using a pinned handle because buffer might be empty and you're trying to avoid special casing it?